File: src/系统/03．技能系统/05．单位技能/01．杂鱼技能/03．第三章/00．封印守卫战公共/01．共享.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__ArraySetLength = ____lualib.__TS__ArraySetLength
local ____exports = {}
local jass = require("jass.common")
local ____require_result_0 = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版")
local stringToFourCCSafe = ____require_result_0.stringToFourCCSafe
local ____require_result_1 = require("lib.扩展函数.BJ函数.02．单位与英雄")
local IsUnitAliveBJ = ____require_result_1.IsUnitAliveBJ
local ____require_result_2 = require("系统.00．核心系统.00．玩家系统.00．英雄注册联动.00．玩家英雄获取桥接")
local getRegisteredPlayerHero = ____require_result_2.getRegisteredPlayerHero
local ____require_result_3 = require("lib.扩展函数.Star扩展函数.02．GS单位属性")
local GS_LoadUintProperty = ____require_result_3.GS_LoadUintProperty
local ____require_result_4 = require("系统.03．技能系统.00．技能模板+函数.02．通用函数.01．控制与Buff")
local _____5355_4F4D_662F_5426_5904_4E8E_786C_63A7_5236_6548_679C_5408_96C6 = ____require_result_4["单位是否处于硬控制效果合集"]
local ____require_result_5 = require("lib.扩展函数.封装函数.01．通用工具.03．特效")
local _____521B_5EFA_70B9_7279_6548 = ____require_result_5["创建点特效"]
local createTimedUnitEffect = ____require_result_5.createTimedUnitEffect
local createUnitEffect = ____require_result_5.createUnitEffect
local destroyUnitEffect = ____require_result_5.destroyUnitEffect
local Player = jass.Player
local GetHandleId = jass.GetHandleId
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local GetUnitFacing = jass.GetUnitFacing
local GetUnitState = jass.GetUnitState
local IssueTargetOrder = jass.IssueTargetOrder
local IssuePointOrder = jass.IssuePointOrder
local IssueImmediateOrder = jass.IssueImmediateOrder
local DestroyEffect = jass.DestroyEffect
local Atan2 = jass.Atan2
local SquareRoot = jass.SquareRoot
local UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE
local _____5F27_5EA6_8F6C_89D2_5EA6 = 57.29577951308232
local _____5141_8BB8_73A9_5BB6_82F1_96C4_6570_91CF = 6
____exports["封印守卫战第三章敌人单位ID"] = {
    ["失控英灵"] = stringToFourCCSafe("n06B"),
    ["夺灵祭司"] = stringToFourCCSafe("n06A"),
    ["锚蚀兽"] = stringToFourCCSafe("n06C"),
    ["断誓猎手"] = stringToFourCCSafe("n06D"),
    ["黑暗残响"] = stringToFourCCSafe("n069"),
    ["裂誓重卫"] = stringToFourCCSafe("n06E"),
    ["失律号令者"] = stringToFourCCSafe("n06F"),
    ["潮蚀巡鳞者"] = stringToFourCCSafe("n056"),
    ["碎礁投石手"] = stringToFourCCSafe("h00Y"),
    ["灵潮祭司"] = stringToFourCCSafe("n054"),
    ["金鳞执刑官"] = stringToFourCCSafe("n052"),
    ["深渊鳞将"] = stringToFourCCSafe("n055")
}
local _____72B6_6001 = {["运行中"] = false, ["敌人列表"] = {}, ["敌人映射"] = {}, ["锚点压制数量"] = {0, 0, 0}}
____exports["设置封印守卫战第三章技能环境"] = function(_____73AF_5883)
    _____72B6_6001["环境"] = _____73AF_5883
    _____72B6_6001["运行中"] = _____73AF_5883 ~= nil
end
____exports["读取封印守卫战第三章技能状态"] = function()
    return _____72B6_6001
end
____exports["封印守卫战单位存活"] = function(unit)
    return unit ~= nil and unit ~= 0 and IsUnitAliveBJ(unit) == true
end
____exports["取封印守卫战单位句柄ID"] = function(unit)
    return unit ~= nil and unit ~= 0 and (GetHandleId(unit) or 0) or 0
end
____exports["创建封印守卫战敌人记录"] = function(unit, _____7C7B_578B, _____5F53_524D_6BEB_79D2)
    local _____53E5_67C4ID = ____exports["取封印守卫战单位句柄ID"](unit)
    if _____53E5_67C4ID == 0 or not ____exports["封印守卫战单位存活"](unit) then
        return nil
    end
    local _____5DF2_6709 = _____72B6_6001["敌人映射"][_____53E5_67C4ID]
    if _____5DF2_6709 ~= nil then
        return _____5DF2_6709
    end
    local record = {
        ["单位"] = unit,
        ["句柄ID"] = _____53E5_67C4ID,
        ["类型"] = _____7C7B_578B,
        ["下次AI毫秒"] = _____5F53_524D_6BEB_79D2,
        ["下次技能毫秒"] = _____5F53_524D_6BEB_79D2,
        ["充能ID"] = 0,
        ["锚点编号"] = 0,
        ["正在压制锚点"] = false,
        ["普攻计数"] = 0,
        ["上次被动毫秒"] = 0,
        ["号令结束毫秒"] = 0,
        ["号令属性已施加"] = false,
        ["号令移动速度增量"] = 0
    }
    local ____72B6_6001__654C_4EBA_5217_8868_6 = _____72B6_6001["敌人列表"]
    ____72B6_6001__654C_4EBA_5217_8868_6[#____72B6_6001__654C_4EBA_5217_8868_6 + 1] = record
    _____72B6_6001["敌人映射"][_____53E5_67C4ID] = record
    return record
end
____exports["读取封印守卫战敌人记录"] = function(unit)
    local id = ____exports["取封印守卫战单位句柄ID"](unit)
    local ____temp_7
    if id > 0 then
        ____temp_7 = _____72B6_6001["敌人映射"][id]
    else
        ____temp_7 = nil
    end
    return ____temp_7
end
____exports["读取封印守卫战敌人列表"] = function()
    return _____72B6_6001["敌人列表"]
end
____exports["移除封印守卫战敌人记录引用"] = function(record)
    if _____72B6_6001["敌人映射"][record["句柄ID"]] == record then
        __TS__Delete(_____72B6_6001["敌人映射"], record["句柄ID"])
    end
    local index = __TS__ArrayIndexOf(_____72B6_6001["敌人列表"], record)
    if index >= 0 then
        __TS__ArraySplice(_____72B6_6001["敌人列表"], index, 1)
    end
end
____exports["清空封印守卫战敌人记录"] = function()
    __TS__ArraySetLength(_____72B6_6001["敌人列表"], 0)
    _____72B6_6001["敌人映射"] = {}
    _____72B6_6001["锚点压制数量"][1] = 0
    _____72B6_6001["锚点压制数量"][2] = 0
    _____72B6_6001["锚点压制数量"][3] = 0
end
____exports["读取封印守卫战核心"] = function()
    local ____opt_8 = _____72B6_6001["环境"]
    local ____temp_10 = ____opt_8 and ____opt_8["读取能量核心"]()
    if ____temp_10 == nil then
        ____temp_10 = nil
    end
    return ____temp_10
end
____exports["读取封印守卫战锚点状态"] = function(_____951A_70B9_7F16_53F7)
    local ____opt_11 = _____72B6_6001["环境"]
    return ____opt_11 and ____opt_11["读取锚点状态"](_____951A_70B9_7F16_53F7)
end
____exports["读取封印守卫战玩家英雄列表"] = function()
    local ____opt_13 = _____72B6_6001["环境"]
    local fromContext = ____opt_13 and ____opt_13["读取玩家英雄列表"]()
    if fromContext ~= nil then
        return fromContext
    end
    local result = {}
    do
        local i = 0
        while i < _____5141_8BB8_73A9_5BB6_82F1_96C4_6570_91CF do
            local hero = getRegisteredPlayerHero(Player(i))
            if ____exports["封印守卫战单位存活"](hero) then
                result[#result + 1] = hero
            end
            i = i + 1
        end
    end
    return result
end
____exports["读取正在修复封印锚点的英雄列表"] = function()
    local ____opt_15 = _____72B6_6001["环境"]
    return ____opt_15 and ____opt_15["读取正在修复锚点的英雄列表"]() or ({})
end
____exports["是封印守卫战玩家英雄"] = function(unit)
    if not ____exports["封印守卫战单位存活"](unit) then
        return false
    end
    local list = ____exports["读取封印守卫战玩家英雄列表"]()
    do
        local i = 0
        while i < #list do
            if list[i + 1] == unit then
                return true
            end
            i = i + 1
        end
    end
    return false
end
____exports["取单位X"] = function(unit)
    return GetUnitX(unit)
end
____exports["取单位Y"] = function(unit)
    return GetUnitY(unit)
end
____exports["取单位面向"] = function(unit)
    return GetUnitFacing(unit)
end
____exports["取两点距离平方"] = function(x1, y1, x2, y2)
    local dx = x2 - x1
    local dy = y2 - y1
    return dx * dx + dy * dy
end
____exports["取单位距离平方"] = function(first, second)
    if not ____exports["封印守卫战单位存活"](first) or not ____exports["封印守卫战单位存活"](second) then
        return 999999999
    end
    return ____exports["取两点距离平方"](
        GetUnitX(first),
        GetUnitY(first),
        GetUnitX(second),
        GetUnitY(second)
    )
end
____exports["取两点距离"] = function(x1, y1, x2, y2)
    return SquareRoot(____exports["取两点距离平方"](x1, y1, x2, y2))
end
____exports["取两点方向角"] = function(x1, y1, x2, y2)
    local angle = Atan2(y2 - y1, x2 - x1) * _____5F27_5EA6_8F6C_89D2_5EA6
    if angle < 0 then
        angle = angle + 360
    end
    return angle
end
____exports["取最近玩家英雄"] = function(unit, _____6700_5927_8303_56F4)
    local list = ____exports["读取封印守卫战玩家英雄列表"]()
    local x = GetUnitX(unit)
    local y = GetUnitY(unit)
    local limit = _____6700_5927_8303_56F4 ~= nil and _____6700_5927_8303_56F4 > 0 and _____6700_5927_8303_56F4 * _____6700_5927_8303_56F4 or 999999999
    local nearest = nil
    local best = limit
    do
        local i = 0
        while i < #list do
            do
                local hero = list[i + 1]
                if not ____exports["封印守卫战单位存活"](hero) then
                    goto __continue39
                end
                local distance = ____exports["取两点距离平方"](
                    x,
                    y,
                    GetUnitX(hero),
                    GetUnitY(hero)
                )
                if distance > best then
                    goto __continue39
                end
                best = distance
                nearest = hero
            end
            ::__continue39::
            i = i + 1
        end
    end
    return nearest
end
____exports["取最近单位"] = function(source, candidates, _____6700_5927_8303_56F4)
    local x = GetUnitX(source)
    local y = GetUnitY(source)
    local limit = _____6700_5927_8303_56F4 ~= nil and _____6700_5927_8303_56F4 > 0 and _____6700_5927_8303_56F4 * _____6700_5927_8303_56F4 or 999999999
    local nearest = nil
    local best = limit
    do
        local i = 0
        while i < #candidates do
            do
                local target = candidates[i + 1]
                if not ____exports["封印守卫战单位存活"](target) then
                    goto __continue44
                end
                local distance = ____exports["取两点距离平方"](
                    x,
                    y,
                    GetUnitX(target),
                    GetUnitY(target)
                )
                if distance > best then
                    goto __continue44
                end
                best = distance
                nearest = target
            end
            ::__continue44::
            i = i + 1
        end
    end
    return nearest
end
____exports["读取单位攻击力"] = function(unit)
    return GS_LoadUintProperty(unit, 2)
end
____exports["读取单位最大生命"] = function(unit)
    return GetUnitState(unit, UNIT_STATE_MAX_LIFE) or 0
end
____exports["读取单位生命"] = function(unit)
    return GetUnitState(unit, jass.UNIT_STATE_LIFE) or 0
end
____exports["单位处于硬控制"] = function(unit)
    return ____exports["封印守卫战单位存活"](unit) and _____5355_4F4D_662F_5426_5904_4E8E_786C_63A7_5236_6548_679C_5408_96C6(unit) == true
end
____exports["命令攻击目标"] = function(unit, target)
    return ____exports["封印守卫战单位存活"](unit) and ____exports["封印守卫战单位存活"](target) and IssueTargetOrder(unit, "attack", target)
end
____exports["命令移动到点"] = function(unit, x, y)
    return ____exports["封印守卫战单位存活"](unit) and IssuePointOrder(unit, "move", x, y)
end
____exports["命令停止"] = function(unit)
    return ____exports["封印守卫战单位存活"](unit) and IssueImmediateOrder(unit, "stop")
end
____exports["播放封印守卫战单位临时特效"] = function(unit, model, duration)
    return createTimedUnitEffect(unit, "origin", model, duration)
end
____exports["创建封印守卫战单位常驻特效"] = function(unit, model, key)
    return createUnitEffect(
        unit,
        "origin",
        model,
        nil,
        key
    )
end
____exports["销毁封印守卫战单位常驻特效"] = function(unit, key)
    destroyUnitEffect(unit, key)
end
____exports["创建封印守卫战点特效"] = function(params)
    return _____521B_5EFA_70B9_7279_6548(params)
end
____exports["销毁封印守卫战特效"] = function(effect)
    if effect ~= nil and effect ~= 0 then
        DestroyEffect(effect)
    end
end
____exports["设置记录锚点压制"] = function(record, enabled)
    if record["锚点编号"] < 1 or record["锚点编号"] > #_____72B6_6001["锚点压制数量"] then
        return
    end
    if record["正在压制锚点"] == enabled then
        return
    end
    local index = record["锚点编号"] - 1
    local before = _____72B6_6001["锚点压制数量"][index + 1] or 0
    local after = enabled and before + 1 or (before > 0 and before - 1 or 0)
    _____72B6_6001["锚点压制数量"][index + 1] = after
    record["正在压制锚点"] = enabled
    if before == 0 ~= (after == 0) then
        local ____opt_17 = _____72B6_6001["环境"]
        if ____opt_17 ~= nil then
            ____opt_17["设置锚点压制"](record["锚点编号"], after > 0)
        end
    end
end
____exports["清理记录锚点压制"] = function(record)
    ____exports["设置记录锚点压制"](record, false)
    ____exports["销毁封印守卫战特效"](record["压制特效"])
    record["压制特效"] = nil
    record["锚点编号"] = 0
end
return ____exports

File: TS/系统/03．技能系统/05．单位技能/01．杂鱼技能/03．第三章/00．封印守卫战公共/01．共享.ts
/** @noSelfInFile */

import type {
  封印守卫战第三章技能环境,
  封印守卫战敌人机制状态,
  封印守卫战敌人记录,
  封印守卫战敌人类型,
  封印守卫战锚点状态,
} from "./00．类型";

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;
const { stringToFourCCSafe } = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版") as {
  stringToFourCCSafe: (this: void, value: string | undefined | null) => number;
};
const { IsUnitAliveBJ } = require("lib.扩展函数.BJ函数.02．单位与英雄") as {
  IsUnitAliveBJ: (this: void, unit: any) => boolean;
};
const { getRegisteredPlayerHero } = require("系统.00．核心系统.00．玩家系统.00．英雄注册联动.00．玩家英雄获取桥接") as {
  getRegisteredPlayerHero: (this: void, player: any) => any;
};
const { GS_LoadUintProperty } = require("lib.扩展函数.Star扩展函数.02．GS单位属性") as {
  GS_LoadUintProperty: (this: void, unit: any, propertyType: number) => number;
};
const { 单位是否处于硬控制效果合集 } = require("系统.03．技能系统.00．技能模板+函数.02．通用函数.01．控制与Buff") as {
  单位是否处于硬控制效果合集: (this: void, unit: any) => boolean;
};
const {
  创建点特效,
  createTimedUnitEffect,
  createUnitEffect,
  destroyUnitEffect,
} = require("lib.扩展函数.封装函数.01．通用工具.03．特效") as {
  创建点特效: (this: void, params: any) => any;
  createTimedUnitEffect: (this: void, unit: any, attachPoint: string, modelPath: string, duration?: number) => any;
  createUnitEffect: (this: void, unit: any, attachPoint: string, modelPath: string, duration?: number, effectKey?: string) => any;
  destroyUnitEffect: (this: void, unit: any, effectKey?: string) => void;
};

const Player = jass.Player as (this: void, playerId: number) => any;
const GetHandleId = jass.GetHandleId as (this: void, handle: any) => number;
const GetUnitX = jass.GetUnitX as (this: void, unit: any) => number;
const GetUnitY = jass.GetUnitY as (this: void, unit: any) => number;
const GetUnitFacing = jass.GetUnitFacing as (this: void, unit: any) => number;
const GetUnitState = jass.GetUnitState as (this: void, unit: any, state: any) => number;
const GetUnitStateJapi = japi.GetUnitState as (this: void, unit: any, state: any) => number;
const IssueTargetOrder = jass.IssueTargetOrder as (this: void, unit: any, order: string, target: any) => boolean;
const IssuePointOrder = jass.IssuePointOrder as (this: void, unit: any, order: string, x: number, y: number) => boolean;
const IssueImmediateOrder = jass.IssueImmediateOrder as (this: void, unit: any, order: string) => boolean;
const DestroyEffect = jass.DestroyEffect as (this: void, effect: any) => void;
const Atan2 = jass.Atan2 as (this: void, y: number, x: number) => number;
const SquareRoot = jass.SquareRoot as (this: void, value: number) => number;
const UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE as any;

const 弧度转角度 = 57.29577951308232;
const 允许玩家英雄数量 = 6;

export const 封印守卫战第三章敌人单位ID = {
  失控英灵: stringToFourCCSafe("n06B"),
  夺灵祭司: stringToFourCCSafe("n06A"),
  锚蚀兽: stringToFourCCSafe("n06C"),
  断誓猎手: stringToFourCCSafe("n06D"),
  黑暗残响: stringToFourCCSafe("n069"),
  裂誓重卫: stringToFourCCSafe("n06E"),
  失律号令者: stringToFourCCSafe("n06F"),
  潮蚀巡鳞者: stringToFourCCSafe("n056"),
  碎礁投石手: stringToFourCCSafe("h00Y"),
  灵潮祭司: stringToFourCCSafe("n054"),
  金鳞执刑官: stringToFourCCSafe("n052"),
  深渊鳞将: stringToFourCCSafe("n055"),
} as const;

const 状态: 封印守卫战敌人机制状态 = {
  运行中: false,
  敌人列表: [],
  敌人映射: {},
  锚点压制数量: [0, 0, 0],
};

export function 设置封印守卫战第三章技能环境(this: void, 环境?: 封印守卫战第三章技能环境): void {
  状态.环境 = 环境;
  状态.运行中 = 环境 != null;
}

export function 读取封印守卫战第三章技能状态(this: void): 封印守卫战敌人机制状态 {
  return 状态;
}

export function 封印守卫战单位存活(this: void, unit: any): boolean {
  return unit != null && unit !== 0 && IsUnitAliveBJ(unit) === true;
}

export function 取封印守卫战单位句柄ID(this: void, unit: any): number {
  return unit != null && unit !== 0 ? (GetHandleId(unit) || 0) : 0;
}

export function 创建封印守卫战敌人记录(
  this: void,
  unit: any,
  类型: 封印守卫战敌人类型,
  当前毫秒: number,
): 封印守卫战敌人记录 | undefined {
  const 句柄ID = 取封印守卫战单位句柄ID(unit);
  if (句柄ID === 0 || !封印守卫战单位存活(unit)) return undefined;
  const 已有 = 状态.敌人映射[句柄ID];
  if (已有 != null) return 已有;
  const record: 封印守卫战敌人记录 = {
    单位: unit,
    句柄ID,
    类型,
    下次AI毫秒: 当前毫秒,
    下次技能毫秒: 当前毫秒,
    充能ID: 0,
    锚点编号: 0,
    正在压制锚点: false,
    普攻计数: 0,
    上次被动毫秒: 0,
    号令结束毫秒: 0,
    号令属性已施加: false,
    号令移动速度增量: 0,
  };
  状态.敌人列表.push(record);
  状态.敌人映射[句柄ID] = record;
  return record;
}

export function 读取封印守卫战敌人记录(this: void, unit: any): 封印守卫战敌人记录 | undefined {
  const id = 取封印守卫战单位句柄ID(unit);
  return id > 0 ? 状态.敌人映射[id] : undefined;
}

export function 读取封印守卫战敌人列表(this: void): 封印守卫战敌人记录[] {
  return 状态.敌人列表;
}

export function 移除封印守卫战敌人记录引用(this: void, record: 封印守卫战敌人记录): void {
  if (状态.敌人映射[record.句柄ID] === record) delete 状态.敌人映射[record.句柄ID];
  const index = 状态.敌人列表.indexOf(record);
  if (index >= 0) 状态.敌人列表.splice(index, 1);
}

export function 清空封印守卫战敌人记录(this: void): void {
  状态.敌人列表.length = 0;
  状态.敌人映射 = {};
  状态.锚点压制数量[0] = 0;
  状态.锚点压制数量[1] = 0;
  状态.锚点压制数量[2] = 0;
}

export function 读取封印守卫战核心(this: void): any {
  return 状态.环境?.读取能量核心() ?? null;
}

export function 读取封印守卫战锚点状态(this: void, 锚点编号: number): 封印守卫战锚点状态 | undefined {
  return 状态.环境?.读取锚点状态(锚点编号);
}

export function 读取封印守卫战玩家英雄列表(this: void): any[] {
  const fromContext = 状态.环境?.读取玩家英雄列表();
  if (fromContext != null) return fromContext;
  const result: any[] = [];
  for (let i = 0; i < 允许玩家英雄数量; i++) {
    const hero = getRegisteredPlayerHero(Player(i));
    if (封印守卫战单位存活(hero)) result.push(hero);
  }
  return result;
}

export function 读取正在修复封印锚点的英雄列表(this: void): any[] {
  return 状态.环境?.读取正在修复锚点的英雄列表() ?? [];
}

export function 是封印守卫战玩家英雄(this: void, unit: any): boolean {
  if (!封印守卫战单位存活(unit)) return false;
  const list = 读取封印守卫战玩家英雄列表();
  for (let i = 0; i < list.length; i++) {
    if (list[i] === unit) return true;
  }
  return false;
}

export function 取单位X(this: void, unit: any): number {
  return GetUnitX(unit);
}

export function 取单位Y(this: void, unit: any): number {
  return GetUnitY(unit);
}

export function 取单位面向(this: void, unit: any): number {
  return GetUnitFacing(unit);
}

export function 取两点距离平方(this: void, x1: number, y1: number, x2: number, y2: number): number {
  const dx = x2 - x1;
  const dy = y2 - y1;
  return dx * dx + dy * dy;
}

export function 取单位距离平方(this: void, first: any, second: any): number {
  if (!封印守卫战单位存活(first) || !封印守卫战单位存活(second)) return 999999999;
  return 取两点距离平方(GetUnitX(first), GetUnitY(first), GetUnitX(second), GetUnitY(second));
}

export function 取两点距离(this: void, x1: number, y1: number, x2: number, y2: number): number {
  return SquareRoot(取两点距离平方(x1, y1, x2, y2));
}

export function 取两点方向角(this: void, x1: number, y1: number, x2: number, y2: number): number {
  let angle = Atan2(y2 - y1, x2 - x1) * 弧度转角度;
  if (angle < 0) angle += 360;
  return angle;
}

export function 取最近玩家英雄(this: void, unit: any, 最大范围?: number): any {
  const list = 读取封印守卫战玩家英雄列表();
  const x = GetUnitX(unit);
  const y = GetUnitY(unit);
  const limit = 最大范围 != null && 最大范围 > 0 ? 最大范围 * 最大范围 : 999999999;
  let nearest: any = null;
  let best = limit;
  for (let i = 0; i < list.length; i++) {
    const hero = list[i];
    if (!封印守卫战单位存活(hero)) continue;
    const distance = 取两点距离平方(x, y, GetUnitX(hero), GetUnitY(hero));
    if (distance > best) continue;
    best = distance;
    nearest = hero;
  }
  return nearest;
}

export function 取最近单位(this: void, source: any, candidates: any[], 最大范围?: number): any {
  const x = GetUnitX(source);
  const y = GetUnitY(source);
  const limit = 最大范围 != null && 最大范围 > 0 ? 最大范围 * 最大范围 : 999999999;
  let nearest: any = null;
  let best = limit;
  for (let i = 0; i < candidates.length; i++) {
    const target = candidates[i];
    if (!封印守卫战单位存活(target)) continue;
    const distance = 取两点距离平方(x, y, GetUnitX(target), GetUnitY(target));
    if (distance > best) continue;
    best = distance;
    nearest = target;
  }
  return nearest;
}

export function 读取单位攻击力(this: void, unit: any): number {
  return GS_LoadUintProperty(unit, 2);
}

export function 读取单位最大生命(this: void, unit: any): number {
  return GetUnitStateJapi(unit, UNIT_STATE_MAX_LIFE) || 0;
}

export function 读取单位生命(this: void, unit: any): number {
  return GetUnitState(unit, jass.UNIT_STATE_LIFE) || 0;
}

export function 单位处于硬控制(this: void, unit: any): boolean {
  return 封印守卫战单位存活(unit) && 单位是否处于硬控制效果合集(unit) === true;
}

export function 命令攻击目标(this: void, unit: any, target: any): boolean {
  return 封印守卫战单位存活(unit) && 封印守卫战单位存活(target) && IssueTargetOrder(unit, "attack", target);
}

export function 命令移动到点(this: void, unit: any, x: number, y: number): boolean {
  return 封印守卫战单位存活(unit) && IssuePointOrder(unit, "move", x, y);
}

export function 命令停止(this: void, unit: any): boolean {
  return 封印守卫战单位存活(unit) && IssueImmediateOrder(unit, "stop");
}

export function 播放封印守卫战单位临时特效(this: void, unit: any, model: string, duration: number): any {
  return createTimedUnitEffect(unit, "origin", model, duration);
}

export function 创建封印守卫战单位常驻特效(this: void, unit: any, model: string, key: string): any {
  return createUnitEffect(unit, "origin", model, undefined, key);
}

export function 销毁封印守卫战单位常驻特效(this: void, unit: any, key: string): void {
  destroyUnitEffect(unit, key);
}

export function 创建封印守卫战点特效(this: void, params: any): any {
  return 创建点特效(params);
}

export function 销毁封印守卫战特效(this: void, effect: any): void {
  if (effect != null && effect !== 0) DestroyEffect(effect);
}

export function 设置记录锚点压制(this: void, record: 封印守卫战敌人记录, enabled: boolean): void {
  if (record.锚点编号 < 1 || record.锚点编号 > 状态.锚点压制数量.length) return;
  if (record.正在压制锚点 === enabled) return;
  const index = record.锚点编号 - 1;
  const before = 状态.锚点压制数量[index] ?? 0;
  const after = enabled ? before + 1 : (before > 0 ? before - 1 : 0);
  状态.锚点压制数量[index] = after;
  record.正在压制锚点 = enabled;
  if ((before === 0) !== (after === 0)) 状态.环境?.设置锚点压制(record.锚点编号, after > 0);
}

export function 清理记录锚点压制(this: void, record: 封印守卫战敌人记录): void {
  设置记录锚点压制(record, false);
  销毁封印守卫战特效(record.压制特效);
  record.压制特效 = undefined;
  record.锚点编号 = 0;
}
